python/tank/authentication/shotgun_shared/saml2_sso.py

```diff
@@ -582,17 +587,150 @@ def on_sso_disable_renewal(self, event):
 ################################################################################
 
 
-def _decode_cookies(app_cookies):
-    sg_cookies = []
-    if app_cookies:
-        sg_cookies = json.loads(base64.b64decode(app_cookies))
-        sg_cookies = [cookie.encode("utf-8") for cookie in sg_cookies]
-    return sg_cookies
+def _decode_cookies(encoded_cookies):
+    """
+    Extract the cookies from a base64 encoded string.
+
+    :param encoded_cookies: An encoded string representing the cookie jar.
+
+    :returns: A SimpleCookie containing all the cookies.
+    """
+    cookies = SimpleCookie()
+    if encoded_cookies:
+        try:
+            decoded_cookies = base64.b64decode(encoded_cookies)
+            cookies.load(decoded_cookies)
+        except TypeError, e:
+            log.error("Unable to decode the cookies: %s" % e.message)
+    return cookies
+
+
+def _encode_cookies(cookies):
+    """
+    Extract the cookies from a base64 encoded string.
+
+    :param cookies: A Cookie.SimpleCookie instance representing the cookie jar.
+
+    :returns: An encoded string representing the cookie jar.
+    """
+    encoded_cookies = base64.b64encode(cookies.output())
+    return encoded_cookies
+
+
+def _get_shotgun_user_id(cookies):
+    """
+    Returns the id of the user in the shotgun instance, based on the cookies.
+
+    :param cookies: A Cookie.SimpleCookie instance representing the cookie jar.
+
+    :returns: A string user id value, or None.
+    """
+    user_id = None
+    for cookie in cookies:
+        if cookie.startswith("shotgun_sso_session_userid_u"):
+            if user_id is not None:
+                # Should we find multiple cookies with the same prefix, it means
+                # that we are using cookies from a multi-session environment. We
+                # have no way to identify the proper user id in the lot.
+                raise Saml2SssoMultiSessionNotSupportedError('Multi-session not supported')
```

Comment:
Would it makse sense to include more info in this error message? `'Multi-session not supported'` feels like a repetition of what we already know from the exception type and isn't super helpful for a client i think. When does this error happen? When you are logged onto multiple sites? Can we include more info in comments and perhaps in the message? How about including site names in the error message. What message would you like to see as a user when you try to log in to toolkit/RV?